<a href="https://colab.research.google.com/github/jrakhshanda/Text-Mining/blob/main/TM_L1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# L1: Information retrieval

In this lab you will apply basic techniques from information retrieval to implement the core of a minimalistic search engine. The data for this lab consists of a collection of app descriptions scraped from the [Google Play Store](https://play.google.com/store/apps?hl=en). From this collection, your search engine should retrieve those apps whose descriptions best match a given query under the vector space model.

## Data set

The app descriptions come in the form of a compressed [JSON](https://en.wikipedia.org/wiki/JSON) file. Start by loading this file into a [Pandas](https://pandas.pydata.org) [DataFrame](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dataframe).

In [1]:
import bz2
import pandas as pd
import warnings
warnings.filterwarnings('ignore', category=DeprecationWarning)
warnings.filterwarnings('ignore', category=FutureWarning)

with bz2.open('app-descriptions.json.bz2') as source:
    df = pd.read_json(source)

In Pandas, a DataFrame is a table with indexed rows and labelled columns of potentially different types. Data in a DataFrame can be accessed in various ways, including by row and by column. To give an example, the code in the next cell shows rows 200–204:

In [2]:
df[200:205]

name                                        description
200  Brick Breaker Star: Space King  Introducing the best Brick Breaker game that e...
201      Brick Classic - Brick Game  Classic Brick Game!\n\nBrick Classic is a popu...
202     Bricks Breaker - Glow Balls  Bricks Breaker - Glow Balls is a addictive and...
203            Bricks Breaker Quest  How to play\n- The ball flies to wherever you ...
204             Brothers in Arms® 3  Fight brave soldiers from around the globe on ...

As you can see, there are two labelled columns: `name` (the name of the app) and `description` (a textual description). The code in the next cell shows how to acess fields from the description column.

In [3]:
df['description'][200:205]

200    Introducing the best Brick Breaker game that e...
201    Classic Brick Game!\n\nBrick Classic is a popu...
202    Bricks Breaker - Glow Balls is a addictive and...
203    How to play\n- The ball flies to wherever you ...
204    Fight brave soldiers from around the globe on ...
Name: description, dtype: object

## Problem 1: Preprocessing

Your first task is to implement a preprocessor for your search engine. In the vector space model, *preprocessing* refers to any kind of transformation that is applied before a text is vectorized. Here you can restrict yourself to a very simple preprocessing: tokenization, stop word removal, and lemmatization.

To implement your preprocessor, you can use [spaCy](https://spacy.io). Make sure that you read the [Linguistic annotations](https://spacy.io/usage/spacy-101#annotations) section of the spaCy&nbsp;101; that section contains all the information that you need for this problem (and more).

Implement your preprocessor by completing the skeleton code in the next cell, adding additional code as you feel necessary.

In [4]:
import spacy
nlp = spacy.load("en_core_web_sm", disable=["tagger", "parser","entityrecognizer"])

In [5]:
def preprocess(text):
  doc = nlp(text)
  res = []
  for token in doc:
    if not token.is_stop:
      if token.is_alpha:
        res.append(token.lemma_)
  return (res)

Your implementation should conform to the following specification:

<strong>preprocess</strong> (<em>text</em>)

> Preprocesses given text by tokenizing it, removing any stop words, replacing each remaining token with its lemma (base form), and discarding all lemmas that contain non-alphabetical characters. Returns the list of remaining lemmas (represented as strings).

**Tip:** To speed up the preprocessing, you can disable loading those spaCy components that you do not need, such as the part-of-speech tagger, parser, and named entity recognizer. See [here](https://spacy.io/usage/processing-pipelines#disabling) for more information about this.

Test your implementation by running the following cell:

In [6]:
preprocess('Apple is looking at buying U.K. startup for $1 billion')

['Apple', 'look', 'buy', 'startup', 'billion']

This should give the following output:


['Apple', 'look', 'buy', 'startup', 'billion']

## Problem 2: Vectorizing

Your next task is to vectorize the data – and more specifically, to map each app description to a tf–idf vector. For this you can use the [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html) class from [scikit-learn](https://scikit-learn.org/stable/). Make sure to specify your preprocessor from the previous problem as the `tokenizer` &ndash; not the `preprocessor`! &ndash; for the vectorizer. (In scikit-learn parlance, the `preprocessor` handles string-level preprocessing.)

In [7]:
import numpy as np
# TODO: Replace the next line with your own code.
#X = np.zeros((len(df), 1))
from sklearn.feature_extraction.text import TfidfVectorizer
#corpus = [preprocess(i) for i in df['description']]
vectorizer = TfidfVectorizer(tokenizer = preprocess)
X = vectorizer.fit_transform( df['description'])

Test your implementation by running the following cell:

In [8]:
X.shape

(1614, 20663)

This should show the dimensions of the matrix `X` to be 1614 × 20669.

## Problem 3: Retrieving

To complete the search engine, your last task is to write a function that returns the most relevant app descriptions for a given query. An easy way to do solve this task is to use scikit-learn&rsquo;s [NearestNeighbors](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html) class. That class implements unsupervised nearest neighbours learning, and allows you to easily find a predefined number of app descriptions whose vector representations are closest to the query vector.

In [12]:
query = 'dodge trains'
from sklearn.neighbors import NearestNeighbors
from sklearn.metrics.pairwise import cosine_distances,cosine_similarity
preprocess(query)
model = NearestNeighbors(n_neighbors=10, metric=cosine_similarity).fit(X)
test = vectorizer.fit_transform([query]).reshape(-1,1)
model.kneighbors(test,return_distance=False)

ValueError: ignored

In [ ]:
query = 'dodge trains'
preprocess(query)

In [ ]:
from sklearn.neighbors import NearestNeighbors
model = NearestNeighbors()
def search(query):
  model = NearestNeighbors()
    # TODO: Replace the next line with your own code.
    return df.iloc[:10]

Your implementation should conform to the following specification:

<strong>search</strong> (<em>query</em>)

> Returns the 10 app descriptions most similar (in terms of cosine similarity) to the given query as a Pandas DataFrame.

Test your implementation by running the following cell:

In [ ]:
search('dodge trains')

The top hit in the list should be *Subway Surfers*.

## Problem 4: Finding terms with low/high idf

Recall that the inverse document frequency (idf) of a term is the lower the more documents from a given collection the term appears in. To get a better understanding for this concept, your next task is to write code to find out which terms have the lowest/highest idf with respect to the app descriptions.

Start by sorting the terms in increasing order of idf, breaking ties by falling back on alphabetic order.

In [ ]:
# TODO: Replace the next line with your own code.
terms = []

Now, print the 10 terms with the lowest/highest idf. How do you explain the results?

In [ ]:
print(terms[:10])

*TODO: Enter your explanation here*

## Problem 5: Keyword extraction

A simple method for extracting salient keywords from a document is to pick the $k$ terms with the highest tf–idf value. Your last task in this lab is to implement this method. More specifically, we ask you to implement a function `keywords` that extracts keywords from a given text.

In [ ]:
def keywords(text, n=10):
    # TODO: Replace the next line with your own code.
    return []

Your implementation should conform to the following specification:

<strong>keywords</strong> (<em>text</em>, <em>n</em> = 10)

> Returns a list with the $n$ (default value: 10) most salient keywords from the specified text, as measured by their tf–idf value relative to the collection of app descriptions.

Test your implementation by running the following cell:

In [ ]:
print(keywords(df['description'][1428]))

This should give the following output:

['train', 'railway', 'railroad', 'rail', 'chaos', 'crash', 'overcast', 'locomotive', 'timetable', 'railyard']

<div class="alert alert-info">
    Please read the General information section on the lab web page before submitting this notebook!
</div>